dash-pipeline/bmv2/stages/tunnel_stage.p4

```diff
@@ -0,0 +1,41 @@
+#ifndef _DASH_STAGE_TUNNEL_P4_
+#define _DASH_STAGE_TUNNEL_P4_
+
+control tunnel_stage(
+    inout headers_t hdr,
+    inout metadata_t meta)
+{
+    action set_tunnel_attrs(
+            @SaiVal[type="sai_ip_address_t"]
+            IPv4Address dip,
+            @SaiVal[type="sai_dash_encapsulation_t", default_value="SAI_DASH_ENCAPSULATION_VXLAN"]
+            dash_encapsulation_t dash_encapsulation,
+            bit<24> tunnel_key) {
+    push_action_static_encap(hdr = hdr,
+                            meta = meta,
+                            encap = dash_encapsulation,
+                            vni = tunnel_key,
+                            underlay_sip = hdr.u0_ipv4.src_addr,
+                            underlay_dip = dip,
+                            overlay_dmac = hdr.u0_ethernet.dst_addr);
+    }
+
+    @SaiTable[name = "dash_tunnel", api = "dash_tunnel", order = 0, isobject="true"]
+    table tunnel {
+        key = {
+            meta.dash_tunnel_id : exact @SaiVal[type="sai_object_id_t"];
+        }
+
+        actions = {
+            set_tunnel_attrs;
+        }
+    }
+
+    apply {
+        if (tunnel.apply().hit) {
+            do_action_static_encap.apply(hdr, meta);
```

Comment:
do you mind to make the tunnel using another encap data instead of the static encap one?

this makes the data plane app hard to implement, because the tunnel info overrides the static encap, which makes us impossible to create the flow.

Reply:
once we have this changed, we can move the tunnel stage to pre_action_apply stage, which makes the pipeline more clear - action apply should be the last thing we do before underlay routing.

Reply:
Decoupled tunnel_encap from static_encap routing action